infra/experimental/sanitizers/ExecSan/execSan.cpp

```diff
@@ -0,0 +1,225 @@
+  debug_log("inspecting");
+  if (path == kTripWire) {
+    kill(pid, SIGKILL);
+    printf("===BUG DETECTED: Shell injection===\n");
+    // TODO: Get/print stacktrace.
+    exit(1);
+  }
+}
+
+void trace(std::set<pid_t> pids) {
+  while (!pids.empty()) {
+    std::vector<pid_t> new_pids;
+
+    auto it = pids.begin();
+    while (it != pids.end()) {
+      auto pid = *it;
+      int status = 0;
+
+      int result = waitpid(pid, &status, __WALL | WNOHANG);
+      if (result == -1) {
+        it = pids.erase(it);
+        continue;
+      }
+
+      if (result == 0) {
+        // Nothing to report yet.
+        ++it;
+        continue;
+      }
+
+      debug_log("finished waiting %d", pid);
+
+      if (WIFEXITED(status) || WIFSIGNALED(status)) {
+        debug_log("%d exited", pid); 
+        it = pids.erase(it);
+        continue;
+      }
+
+      bool is_syscall = WIFSTOPPED(status) && WSTOPSIG(status) == (SIGTRAP | 0x80);
+      int sig = 0;
+      if (!is_syscall) {
+        // Handle generic signal.
+        siginfo_t siginfo;
+        if (ptrace(PTRACE_GETSIGINFO, pid, nullptr, &siginfo) == -1) {
+          debug_log("ptrace(PTRACE_GETSIGINFO, %d): %s", pid, strerror(errno));
+          continue;
+        }
+        sig = siginfo.si_signo;
+        debug_log("forwarding signal %d to %d", sig, pid);
+      }
+
+      if (WIFSTOPPED(status) && 
+          (status>>8 == (SIGTRAP | (PTRACE_EVENT_CLONE<<8)) ||
+           status>>8 == (SIGTRAP | (PTRACE_EVENT_FORK<<8)) ||
+           status>>8 == (SIGTRAP | (PTRACE_EVENT_VFORK<<8)))) {
+        long new_pid;
+        if (ptrace(PTRACE_GETEVENTMSG, pid, 0, &new_pid) == -1) {
+          debug_log("ptrace(PTRACE_GETEVENTMSG, %d): %s", pid, strerror(errno));
+          continue;
+        }
+        debug_log("forked %ld", new_pid);
+        new_pids.push_back(new_pid);
+      }
+
+      if (is_syscall) {
+        // TODO: distinguish between syscall enter and exit.
+        user_regs_struct regs;
+        if (ptrace(PTRACE_GETREGS, pid, 0, &regs) == -1) {
+          debug_log("ptrace(PTRACE_GETREGS, %d): %s", pid, strerror(errno));
+          continue;
+        }
+
+        if (regs.orig_rax == __NR_execve) {
+          inspect(pid, regs);
+        }
+        // TODO: Check for commands with invalid syntax passed to /bin/sh and
+        // other shells.
+      }
+
+      debug_log("tracing %d %d", pid, sig);
+      if (ptrace(PTRACE_SYSCALL, pid, nullptr, sig) == -1) {
+        debug_log("ptrace(PTRACE_SYSCALL, %d): %s", pid, strerror(errno));
+        continue;
+      }
+
+      ++it;
+    }
+
+    pids.insert(new_pids.begin(), new_pids.end());
+  }
+}
+
+int main(int argc, char **argv) {
+  if (argc <= 1) {
+    fatal_log("Expecting at least one arguments, received %d", argc - 1);
+  }
+
+  // Create an executable tripwire file, as programs may check for existence
+  // before actually calling exec.
+  std::ofstream tripwire(kTripWire);
+  tripwire.close();
```

Comment:
I guess a 0 size executable isn't so realistic but I don't have better idea (write the world's tiniest ELF?)

Reply:
ack, we can change this later if this becomes a problem.